source/slang/slang-ir-legalize-varying-params.cpp

```diff
@@ -0,0 +1,1321 @@
+
+
+    void processEntryPoint(IRFunc* entryPointFunc, IREntryPointDecoration* entryPointDecor)
+    {
+        m_entryPointFunc = entryPointFunc;
+
+        // Before diving into the work of processing an entry point, we start by
+        // extracting a bunch of information about the entry point that will
+        // be useful to the downstream logic.
+        //
+        m_stage = entryPointDecor->getProfile().getStage();
+        m_firstBlock = entryPointFunc->getFirstBlock();
+        m_firstOrdinaryInst = m_firstBlock ? m_firstBlock->getFirstOrdinaryInst() : nullptr;
+
+        auto entryPointLayoutDecoration = entryPointFunc->findDecoration<IRLayoutDecoration>();
+        SLANG_ASSERT(entryPointLayoutDecoration);
+
+        auto entryPointLayout = as<IREntryPointLayout>(entryPointLayoutDecoration->getLayout());
+        SLANG_ASSERT(entryPointLayout);
+
+        // Note: Of particular importance is that we extract the first/last parameters
+        // of the function *before* we allow the subtype to perform per-entry-point
+        // setup operations. This ensures that if the subtype adds new parameters to
+        // the beginnign or end of the parameter list, those new parameters won't
+        // be processed.
+        //
+        IRParam* firstOriginalParam = m_firstBlock ? m_firstBlock->getFirstParam() : nullptr;
+        IRParam* lastOriginalParam = m_firstBlock ? m_firstBlock->getLastParam() : nullptr;
+
+        // We allow the subtype to perform whatever setup or code generation
+        // it wants to on a per-entry-point basis. In some cases this might
+        // inject code into the start of the function to provide the value
+        // of certain system-value parameters.
+        //
+        beginEntryPointImpl();
+
+        // We now proceed to the meat of the work.
+        //
+        // We start by considering the result of the entry point function
+        // if it is non-`void`.
+        //
+        auto resultType = entryPointFunc->getResultType();
+        if( !as<IRVoidType>(resultType) )
+        {
+            // We need to translate the existing function result type
+            // into zero or more varying parameters that are legal for
+            // the target. An entry point function result should be
+            // processed in a way that semantically matches an `out` parameter.
+            //
+            auto legalResult = createLegalVaryingVal(
+                resultType,
+                entryPointLayout->getResultLayout(),
+                LayoutResourceKind::VaryingOutput);
+
+            // Now that we have a representation of the value(s) that will
+            // be used to hold the entry-point result we need to transform
+            // any `returnVal(r)` instructions in the function body to
+            // instead assign `r` to `legalResult` and then `returnVoid`.
+            //
+            IRBuilder builder(m_sharedBuilder);
+            for( auto block : entryPointFunc->getBlocks() )
```

Comment:
Does `getBlocks` function allows you to recursively traverse all blocks in a function? After reading the implementation of `IRInstList` I still don't see why that is the case.